### PR TITLE
feat(operator/operator-chart)!: add tuple generator CRG and support for tuple-type-specific thresholds and priorities

### PR DESCRIPTION
**Pull request by strieflin**

### CODE REVIEW

**Comment by strieflin** [comment]
There is a related PR carbynestack/carbynestack#74 that updates the deployment logic to use this version of Klyshko.

File: klyshko-operator/controllers/tuplegenerationscheduler_controller.go [hunk 1]
```diff
@@ -103,10 +124,61 @@ func (r *TupleGenerationSchedulerReconciler) Reconcile(ctx context.Context, req
...
+				generatorsByTupleType[support.Type] = []klyshkov1alpha1.TupleGenerator{generator}
+			}
+		}
+	}
+	return generatorsByTupleType, nil
+}
+
+// getServiceablePolicies filters the policies declared on the given scheduler resource by removing those policies for
+// which no generator is available in the given map of generators
+func (r *TupleGenerationSchedulerReconciler) getServiceablePolicies(
```
Comment (sbckr) [inline klyshko-operator/controllers/tuplegenerationscheduler_controller.go:149]: Update comment to reflect that policies are not considered if more than one generator is available for a given tuple type

**Review by sbckr — CHANGES_REQUESTED** [review]
Please have a look at individual comments

File: README.md [hunk 2]
```diff
@@ -114,6 +114,47 @@ and secrets (see [here](#configuration-parameters) for details). Consult the
...
+      batchSize: 100000
+    - type: INPUT_MASK_GF2N
+      batchSize: 100000
+    - type: INVERSE_TUPLE_GF2N
+      batchSize: 100000
+    - type: SQUARE_TUPLE_GF2N
+      batchSize: 100000
+    - type: MULTIPLICATION_TRIPLE_GF2N
+      batchSize: 100000
+```
```
[the remark this answers is not in the document]
Comment (strieflin) [reply README.md:150]: Assuming you mean `kubectl` here? "Implemented" the second approach.

File: README.md [hunk 3]
```diff
@@ -114,6 +114,47 @@ and secrets (see [here](#configuration-parameters) for details). Consult the
...
+
+This registers the generator with Klyshko. Note that you have to specify each
+tuple type supported by the CRG and provide a recommended batch size for jobs
+that generate that type of tuples. In case a tuple type is supported by multiple
+generators no tuples are generated for that tuple type to avoid potential
+inconsistencies across VCPs.
```
[the remark this answers is not in the document]
Comment (strieflin) [reply README.md:156]: Converted that into a more prominent **IMPORTANT** section. Validation is possible by means of admission webhooks. I have created #83 to track this idea.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (sbckr) [reply README.md:150]: 😅

File: klyshko-operator/api/v1alpha1/tuplegenerator_types.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+// TupleGeneratorStatus defines the observed state of TupleGenerator.
+type TupleGeneratorStatus struct {
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:resource:shortName=tg;tgenerator
+//+kubebuilder:subresource:status
+
+// TupleGenerator is the Schema for the tuplegenerators API.
```
[the remark this answers is not in the document]
Comment (strieflin) [reply klyshko-operator/api/v1alpha1/tuplegenerator_types.go:55]: Done

File: klyshko-operator/charts/klyshko/templates/crds.yaml [hunk 5]
```diff
@@ -125,48 +125,50 @@ metadata:
...
+      - tg
+      - tgenerator
+    singular: tuplegenerator
   scope: Namespaced
   versions:
     - name: v1alpha1
       schema:
         openAPIV3Schema:
-          description: TupleGenerationScheduler is the Schema for the TupleGenerationScheduler API.
+          description: TupleGenerator is the Schema for the tuplegenerators API.
```
[the remark this answers is not in the document]
Comment (strieflin) [reply klyshko-operator/charts/klyshko/templates/crds.yaml:144]: Done

File: klyshko-operator/controllers/config_utils_test.go
[the remark this answers is not in the document]
Comment (strieflin) [reply klyshko-operator/controllers/config_utils_test.go:1]: Done

File: klyshko-operator/controllers/tuplegenerationscheduler_controller.go (hunk 1 above)
Comment (strieflin) [reply klyshko-operator/controllers/tuplegenerationscheduler_controller.go:149]: Done

**Review by sbckr — APPROVED** [review]
LGTM

**Pull request merged**